Pan effect

jclaveau opened the pull request:
I tried to reproduce a Pan effect I have on my mixer which simply bounces the sound from a side to the other. It presently looks almost the same despite the period is a multiple of the bpm on my hardware.

Three parameters are set : 
- Depth
- Period
- Strength (probably not the good name) which defines the easing of the sound when it moves from a speaker to the other.

There a still some questions I'm not able to answer which are explained in comments.

I don't know if it will be considered as mergeable but it was, a least, a really good practice for a first effect :)

demos committed:
almost working

demos committed:
steps working

demos committed:
forgotten files

demos committed:
cleaning

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect

demos committed:
all controls implemented

demos committed:
beginning ramping

demos committed:
position changed each 2 samples

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect

demos committed:
ramping + cleaning

demos committed:
Revert "ramping + cleaning"

This reverts commit 92aee711c1705fe50174d3328f34dfb6030975f5.

demos committed:
ramping + cleaning

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
not working

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
cleaning

demos committed:
comments and review

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
missing header for windows

daschuer commented:
After a quick overview, your code looks good, thank you. 
It is possible to make it synced with the beats, 
you may have a look at https://github.com/mixxxdj/mixxx/pull/337
This branch uses beat sync. It could be also interesting to introduce a similar approach for the controls, like proposed here https://github.com/pwhelan/mixxx/pull/1#issuecomment-59495856

* Button: Periods [1| 2 |4 | 16 | Tempo]
* Knob: Tempo 0 .. 200 Bpm 

The other two are not self explaining, however I do not now good names for this:   
Probably: Strange = curve?  

It might hard to tweak all of them live anyway. Do you really need both, or is it possible to combine them for a cool effect? 

Name: 
Fo

daschuer commented:
Name: 
For me, a pan effect is a simple knob labeled with "Pan" where you can orientate a mono signal in the stereo output. I like to reserve the name for this.
Do you like to implement it in a separate PR?

This on can be renamed to "PingPong" or "Bee"

You can improve the effect a lot, if you add a delay for the reduced channel (left or right). 
Such a delay line is also required for the "Pan" effect otherwise it will be a "Balance" Effect. 

You may reuse https://github.com/mixxxdj/mixxx/blob/master/src/engine/enginefilterdelay.h
or https://github.com/mixxxdj/mixxx/blob/master/src/engine/enginedelay.cpp

It is no problem, if you like to merge this without the advanced feature though.

egasimus commented:
PingPong evokes a delay. Why not just AutoPan?

daschuer commented:
Ah, yes. This seams to be the most common term for this. 

I have just found http://www.tonehome.de/ibanez-effects/big-box-series/fp-777-flying-pan/
It is a auto panner combined with a pahser. Interesting.

demos committed:
renaming as AutoPan

demos committed:
available as chain

demos committed:
period as multiple of beat duration

jclaveau commented:
@daschuer , @egasimus thanks a lot for your review, advices and links which saved a lot of my time.

+ I renamed this effect as AutoPan
+ I added the period as a multiple of the beat duration (from 1/8 to 64 for testing)
+ AutoPan is now testable in LateNight

My remaining questions : 
Is there a way to change the behaviour of the period button so it would have stops on each of the 9 possible positions when the beat length is available (like the knob for setting loop length on traktor's controllers, which is a pushbutton too btw)? Furthermore, the default value for this button is not the same depending of this configuration. 
If saw you spoke about buttons for the configuration of the goaslicer which looks really interesting. However, I think I will try to implement the delay first to have all the parameters in the hands before testing interfaces.


Concerning the effect itself, at the beginning I used SampleUtil::mixStereoToMono() to make a real pan. The result was an audible changing of the sound that wasn't pleasent so I checked on my mixer and I realised they had called pan what, in fact, is a balance (if i'm not misunderstanding). So I decided to simply mimic the hardware. I think the pan wouldn't be hard to implement with a ramping during the change from stereo to mono.
Otherwise, AutoBalance is still possible :P 


Regarding the code, I wonder if there is a folder to store tools for effects development (in my case the sample that prevents clicks automatically or, why not, the sqare to sinusoid curve).


Anyway, this effect is already quite nice to use but I'd like to try to implement the delay (just for learning and if it doesn't block it for shipping in 1.12).

daschuer commented:
> Is there a way to change the behaviour of the period button so it would have stops on each of the 9 possible positions when the beat length is available (like the knob for setting loop length on traktor's controllers, which is a pushbutton too btw)? 

Mixxx features a button control with a flyout menu on the right mouse button CONTROL_TOGGLE_STEPPING
You can use EffectManifestParameter::appendStep() 
It is used here https://github.com/mixxxdj/mixxx/blob/957a283fc8ce84565b246218999f0a0f3b8fd10c/src/effects/native/bessel4lvmixeqeffect.cpp

>  I used SampleUtil::mixStereoToMono()  ....

Ah, sorry for the misunderstanding. I did not mean to Mixxx down the signal to mono for pan. I was referring a mono Input like a Mic, that is placed in the stereo panorama by a pan pot.  
But it sound like an interesting approach. 

For me a pure mono input pan effect tweaked to right, is something like this:
```
               |------ -Gain -- Delay --- L
--Mic (Mono)---|
               |------------------------- R
```
Mm, our tracks normally have already a stereo panorama. How van we tweak this? 

This will move the orientation of the listener position in the existing panorama and is what IMHO suits for a simple "Pan" effect: 
```
-- Track L ------ -Gain -- Delay --- L

-- Track R ------------------------- R
```
 
What will a mono downmix do? Probably like that?  Change the distance of the listener position? 
```
-- Track L --\-+ -Gain -- Delay --- L
              X
-- Track R --/-+------------------- R
```

> Regarding the code, I wonder if there is a folder to store tools for effects development (in my case the sample that prevents clicks automatically or, why not, the sqare to sinusoid curve).

I think you have already found /src/util folder. Bigger helpers, processing samples, can be placed in /src/engine.

daschuer commented on src/effects/native/autopaneffect.h:
```diff
@@ -0,0 +1,104 @@
+        maxDifference = newMaxDifference;
+    }
+    
+    inline RampedSample & operator=(const float newValue) {
```
This is a nice trick to hide functions from a reviewer ;-)

I would prefer not to override =, instead find a function name that reflects what it does.

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,196 @@
+        }
+        
+        // transform the position into a sinusoid (but between 0 and 1)
+        gs.frac = (sin(M_PI * 2.0f * position) + 1.0f) / 2.0f;
```
Mm, I do not understand how this will work in any case. Probably there are still click issues. 
If clicking occurs during parameter change, we have nice functions that fade from on value to an other used in the other effects.

jclaveau replied:
@daschuer  Actually, there is no clicks anymore glad to the RampedSample class.

Here is a schema to explain the relation between the position and the "frac" value : 


![pan_effect_frac_curves](https://cloud.githubusercontent.com/assets/1556489/5806979/a89bed06-a01e-11e4-9f80-8e27cadffb62.png)

The steps are in red and the "slope" is corresponding to the parts that are not steps in the position.

From the beginning to the end, you change the size of the steps from 0 to a half of the period (strength parameter).

What you can see here is that the clicks don't come from the parameter itself but from the slope (a square curve or a too short slope would have made clicks before).

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

daschuer commented:
Thank you for you drawings. To be honest, I do not fully understand them. How are the x and y axis labeled? 
I have just tested the effect and it is hard to hear a difference, when turning the Strange and Ramping knob. However I see the effect in the VU meters. This might be improved once we have the delay, but it seams to be a sign to combine some parameters. 
The effect scales with the audio buffer size. This should not happen.
IMHO this effect code could be simplified, If you calc the two gain values first and then ramp from the old gains to the just calculated gains using copyWithRampingGain(). This is a established behaviour in Mixxx to prevent clicks and this will separate the effect code from the click prevention code.

jclaveau commented:
@daschuer 
Sorry, I was hurry when I begun the schema so here is a detailed version of those curves. Please, tell me if there is still something hard to understand.  



![pan_effect_frac_curves](https://cloud.githubusercontent.com/assets/1556489/5887459/1c0d3432-a3d2-11e4-9881-b848fe62115a.png)



Concerning the "Strength" parameter, I cannot agree with you when you say it has not a big impact.
+ At its maximum value, the sound is cut on a side and fully enabled on the other. It can never be "in the middle".
+ At its minimum value, the sound goes smoothly from a channel to another and never "stops" on a side.
+ Between those two values, the sound will stays on a side for the duration of the "step" then smoothly move to the other side where it will stays again for the duration of the steps etc.

Concerning the ramping one, it has never been made to stay (as the comments explain it) so your report is absolutely normal. 

So, now the drawing is more explicit, I have to : 
+ Rename "Strength" as "Curve" so you will stop to call it "Strange" parameter ;)
+ Replace my ramping mechanism by copyWithRampingGain (so the ramping threshold parameter will not be a question anymore and the code simplified).
+ Add the delay you talked about.
+ Fix the scaling with the buffer size.

Thanks, a lot for the reports!

daschuer commented:
Ah, yes i think I got it it. 
Is the "faction" is the position in the stereo panorama, that finally drives the pan knob?  
You may rename "position" to avoid future confusion. 

Your plan sounds good. I am curious how it will sound with the delay. 

Your depth knob seams to be the same what the external mix (dry/wet) knob does. Is it required to have it inside the effect? Is the full amplitude sinus curve on one side of the "curve" knob? Maybe you can combine it with the depth knob and allow to select sinus curves with a smaller amplitudes by the curve knob. 

I have just found this http://www.residentadvisor.net/feature.aspx?1838
including a nice demo of an auto pan effect. 
And http://www.soundonsound.com/sos/sep10/articles/pt-0910.htm 
Not sure, but maybe these links are helpful.

demos committed:
cleaning + remove buffersize from period count

demos committed:
removing useless depth + renaming position as angleFraction + setting good ramping threshold value

jclaveau commented:
@daschuer 
Mmm thanks a lot for the documentation, I didn't know that Resident Advisor did such articles! I will read that carefully when I'll have time.

+ I removed the depth and ramping parameters.
+ Renamed the Strength as Curve and position as angleFraction (please tell me if it's enough meaningful for you).
+ The = operator for the RampedSample is now replaced by setWithRampingApplied(). 
+ I tried to use copyWithRampingGain but realized the ramping is done on a full buffer only. As my fraction computing is done for each sample, I kept my live ramping mechanism (based on a maximum difference for each value change) so the ramping is applied even on multiple processGroup calls if needed.

Next mission : 
+ Read the docs you gave be to ask less dummy questions and become less dummy at all. :p
+ add the delay and a parameter to control it and test the result.
+ Find how to make a real pan effect and not a balance one (maybe rename it as AutoBalance if the result is not better for the ears).

Thanks a lot for your quick answer.

demos committed:
Experiments with the delay

demos committed:
master merge + conflict

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
merging conflict with samplutil optimisations

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
changing group to channel

demos committed:
Renaming Curve parameter as Smoothing parameter as in  SoundToys PanMan plug-in

demos committed:
delay + period has now seconds as unit if beat_length is unavailable + cleaning

jclaveau commented:
@daschuer 

I resolved the conflicts with master and made some improvements as you told me.
+ I finally found a good name for the curve parameter in the article you shown me : Smoothing which is used in the SoundToys PanMan plugin (http://fr.audiofanzine.com/plugin-traitement-spatial/soundtoys/panman/avis/)
+ I added a delay on the channel having the reducing gain (with a bounded parameter to test different values).
+ I corrected the period parameter so it's now calculated in beats (if beat_length is available) or in seconds otherwise (for master channel for example). 

I noticed that the delay produces clicks, especially if the period is short. I don't understand the reason apart I'm sure it's not due to the smoothing parameter. I probably misuse it.

Some improvements can be done as adding an offset when the smoothing is hard mainly (to avoid jumps from a channel to another during the kick). I'll give a look another day.

Thanks in advance for your feedback.

daschuer commented:
This PR includes unrelated skin files

daschuer commented on src/sampleutil.h:
```diff
@@ -135,7 +135,7 @@ class SampleUtil {
     static void copyWithRampingGain(CSAMPLE* pDest, const CSAMPLE* pSrc,
             CSAMPLE_GAIN old_gain, CSAMPLE_GAIN new_gain,
-            int iNumSamples);
+            int iNumSamples, bool left=true, bool right=true);
```
It looks like this change is unused.

jclaveau replied:
Removed

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,192 @@
+    gs.frac.ramped = false;     // just for debug
+    
+    // prepare the delay buffer
+    float delay = round(m_pDelayParameter->value() * sampleRate);
```
The delay should change with the pan position. 
I think it might be possible th pick are reasonable delay and do not need a parameter for this.

jclaveau replied:
The parameter is only here to challenge the delay value, I totally agree with you.
I think the offset can have a fixed value too (probably changing with the period) and finally the width parameter could have a dedicated knob (i'm speacking about the parameters provided in the SoundToys PanMan plugin for Protools).

demos committed:
cleaning unrelated skin files

demos committed:
removing unused change

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

daschuer commented:
Just tested the auto pan, and it works nice. 
Here some Ideas for improvements. 
* Alternating delay 
* A Pushbutton to select the Period relative to the beats, in Addition to the Period knob. 
* Add a second half to the Smoothing parameter that allows to center the effect, in a way that it not reaches fully left and fully right position. Idea: Smoothing = 0, disables the effect. 
Once we have the delay, the dry wet control will not work for this.

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
tests with the delay

demos committed:
width parameter + not working tests with delay

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

jclaveau commented:
+ I have added the width parameter. When you say "Add a second half to the Smoothing parameter", do you mean a special UI?
+ I tried a dozen of ways to use the EngineFilterDelay as you told me but I get nothing more than clics. I think I need help here. :/

What I still have to do : 
+ The pushbutton to choose between a period in seconds and a period in beats.
+ Change the lawCoef to have a really dry effect when width is at 0.
+ The delay but I'm stuck.

demos committed:
cleaning + removing useless lawCoef

demos committed:
cleaning

demos committed:
cleaning

demos committed:
cleaning

daschuer committed:
Introduced a EngineFilterPan class

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
Merge branch 'pan_effect_stereo' of https://github.com/daschuer/mixxx into pan_effect_stereo

demos committed:
reenabling pan curve

demos committed:
Merge pull request #1 from daschuer/pan_effect_stereo

Introduced a EngineFilterPan class
Thank you very much, i'll try to integrate it fastly :)

rryan commented:
What else needs to be done here?

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
missing coef

demos committed:
Merge branch 'pan_effect_stereo' of https://github.com/demos/mixxx into pan_effect_stereo

demos committed:
beginning of period unit switch

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
the delay moves with the sound

demos committed:
cleaning

jclaveau commented:
@rryan sorry for the late response.

Presently, the sound goes from left to wright without clicks (at least on my laptop).
We can choose : 
+ If it goes softly or hardly from a side to an other with the "Smoothing" parameter.
+ If it goes fully or partially from a side to another with the "Width" parameter.
+ The period is based on the bpm or a given time (with the "period unit" toggler and the "period" knob")

> What else needs to be done here?
It's mainly a question of UX now and, as I have no effect controller and my mixer allows only the control of the period, I have no idea how to make something easy to use.
+ The superknob could control the smoothing and width params together for example
+ I think I should use CONTROL_KNOB_LOGARITHMIC for the period knob and try to define a similar scale when it's defined in beats or in seconds.

@daschuer 
+ If I understood well your pan delay, it's now correctly used (the delay is defined depending on the position of the sound. Doing this, if the width param is at 0, the delay will be stuck at 0 two and the effect will be dry).
+ The pushbutton to enable beats as unit for the period works (if the bpm is available obviously).
+ I noticed that this pushbutton doesn't exist on LateNight. I wonder if it's a bug as there is no other effect using it but equalizers effects.

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+    delay->setMinimum(-0.01);
+    delay->setMaximum(0.01);
+    delay->setDefault(0.0);
+    /**/
```
compiler warning: "/*" within comment

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+    // Width : applied on the channel with gain reducing.
+    EffectManifestParameter* width = manifest.addParameter();
+    width->setId("width");
+    width->setName(QObject::tr("width"));
```
Width

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+    period->setName(QObject::tr("Period"));
+    period->setDescription("Controls the speed of the effect.");
+    period->setControlHint(EffectManifestParameter::CONTROL_KNOB_LINEAR);
+//    period->setControlHint(EffectManifestParameter::CONTROL_KNOB_LOGARITHMIC);
```
I like the logarithmic version more.

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+//    period->setControlHint(EffectManifestParameter::CONTROL_KNOB_LOGARITHMIC);
+    period->setSemanticHint(EffectManifestParameter::SEMANTIC_UNKNOWN);
+    period->setUnitsHint(EffectManifestParameter::UNITS_UNKNOWN);
+    period->setMinimum(0.01);
```
This can be smaller

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+    // Period unit
+    EffectManifestParameter* periodUnit = manifest.addParameter();
+    periodUnit->setId("periodUnit");
+    periodUnit->setName(QObject::tr("Period Unit"));
```
I think this can be called "Beat Sync" or just "sync" or "Beats"

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+        double beats = pow(2, floor(period * 9 / m_pPeriodParameter->maximum()) - 3);
+        period = groupFeatures.beat_length * beats;
+    } else {
+        // max period is 50 seconds
```
50 seconds: Still correct?

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+    CSAMPLE period = m_pPeriodParameter->value();
+    if (periodUnit == 1 && groupFeatures.has_beat_length) {
+        // 1/8, 1/4, 1/2, 1, 2, 4, 8, 16, 32, 64
+        double beats = pow(2, floor(period * 9 / m_pPeriodParameter->maximum()) - 3);
```
needs to be tweaked for log scale knob.

daschuer replied:
pow() is slow. Use an if else tree, or a lookup table.

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+    // apply the delay
+//    float delay = round(m_pDelayParameter->value() * sampleRate);
+//    gs.delay->setLeftDelay( delay * sinusoid );
+    gs.delay->setLeftDelay( 0.01 * sinusoid );
```
This does not work. I have just played with it a bit and it turns out that
```
(-0.005 * math_clamp(sinusoid, -1.0, 1.0) * sampleRate) 
```
does a nice job

daschuer commented on src/effects/native/autopaneffect.cpp:
```diff
@@ -0,0 +1,218 @@
+    smoothing->setName(QObject::tr("Smoothing"));
+    smoothing->setDescription(
+            QObject::tr("How fast the signal goes from a channel to an other"));
+    smoothing->setControlHint(EffectManifestParameter::CONTROL_KNOB_LINEAR);
```
logarithmic works better for me.

daschuer commented:
Thank you @jclaveau. Nice.

I have left some comments inline.
The knob layout works for me like it is. The missing knob in LateNight is a separate bug. 
If you think we have to improve the UI, we may consider to use a multistate button to click though the fractions of beats. These type of buttons have a right click menu to select a dedicated value. 
Use periodUnit->appendStep(..) to add steps. 
They are required for LV2 effects and already used in the LV2 branch: 
https://github.com/mixxxdj/mixxx/pull/359

daschuer commented:
The parameter descriptions needs also some love. Since I am not a English native, I cannot help much here.

daschuer commented:
@rryan: How do we handle such PR for 1.12. Is it still desired to merge something like this to 1.12 or should this go to master? 
Will we have an official "String Freeze" or do we treat the beta as this?

ywwg commented:
Since we plan to do releases more often I would say this should target master.  The beta branch should just be for fixing problems in existing code.

demos committed:
cleaning + better texts + delay debug

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
changing the scale of the period knob

demos committed:
changing the scale of the period knob

demos committed:
logarithmic smoothing + cleaning

jclaveau commented:
@daschuer Thank you for the advices and review. All your comments are addressed now.
@rryan @ywwg I'm not a english native neither so your feedback about the wording is welcome.

> If you think we have to improve the UI, we may consider to use a multistate button to click though the fractions of beats.

I really don't understand how it could be better to choose the beats fraction via the contextual menu (it would be a terribly slow movement if you don't have a controller). More, the period knob changes the beats fraction when beats unit are enabled.

Let me know if you want me to improve something else.

jclaveau commented:
Notice : the tests pass on my laptop.

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

daschuer committed:
enginefilterpane: use linear stetching instead of ramping

daschuer committed:
add enginefilterpanesingle

daschuer committed:
Used enginefilterpanesingle in auto pane effect

daschuer commented:
The knob behavior is nice now. Thank you. 

It turns out that the delay and the gain processing needs to be in sync to avoid clicks, for the case smothing at max. I currently am preparing EngineFilterPanSingle as an PR to this branch, that can be placed into the sample loop.  

The sync code seams to fail for curtained tracks. I have not investigates it enough to have a conclusion. I expect that if this works correct, it should be possible to have the "boom" left and the "its" right. 
Currently it is moving around even though "sync" is enabled. It probably depends on my tracks analyses with a non const beat grid. Does it work for you?

daschuer committed:
Use ramped value for delay

daschuer commented:
Syncing works, if the track has constant beat grid. So you probably have to deal with the fraction of a beat as well to avoid cycling. But since this is an advanced issue, we can postpone it and just file a bug report. 

It would be nice if the period knob has a "Infinity" value at its max position. This way we are able to set the panning at any position. 

Other than that, the effect work really great and its a lot of fun. I am curious to test is on a big sound system. Thank you for all the work.

daschuer commented:
You may have a look at the syncing coder of the Goaslicer.
https://github.com/mixxxdj/mixxx/pull/337/files

demos committed:
Merge branch 'master' of git://github.com/mixxxdj/mixxx into pan_effect_stereo

demos committed:
adding comments to filterpansingle

demos committed:
adding comments to filterpansingle 2

demos committed:
Merge pull request #2 from daschuer/pan_effect_stereo

pan delay per sample with linear stretching

demos committed:
Merge branch 'daschuer_filterpansingle_for_autopan' of https://github.com/demos/mixxx into pan_effect_stereo

demos committed:
Merge branch 'pan_effect_stereo' of https://github.com/demos/mixxx into pan_effect_stereo

demos committed:
pausable period

demos committed:
position of the sound fixed when period knob is used

demos committed:
period will not be fixed anymore if the effect is in enabling state or the position change is ramped

demos committed:
period will not be fixed anymore if the effect is in enabling state or the position change is ramped

demos committed:
reintroducing lawcoef : *2 at center and * 1.7 at left or right

demos commented:
@daschuer, 

+ I have added the support for an infinite period. Changing the period length will no longer move the spatial position of the sound.
+ I re-added the pan law support but without bug : *2 at the center and *1 + 1 / sqrt(2) on the edges.
+ When you tell me to look at the goaslicer sync, do you mean I should try to implement the quantization here too? I decided to don't do that to let people choose if they want the bass at the center or on the edges, maybe I'm wrong.

It looks close to the end now :), thanks a lot for your help!

daschuer commented:
Yes, I have just tested it and it sounds great. 
The code is also OK. Thank you very much.

demos commented:
Thanks a lot for all your all your help and merging.
I played a bit with it too and I found it really funny when applied on two tracks or when combined with an EQ (increased basses with the autopan having a period of 1/16 creates a heavy dirty sound).

Thanks again! :)
